Spring webflow: redirect url contains String[] parameter values

If I specify the attribute `redirect="true"` on a view in one of the webflows (`login-webflow.xml` for example) I end up with serialized `String[]` in the URL. So the URL would look like:

`/login?client_name=%5BLjava.lang.String%3B%40221ebc09&oauth_token=%5BLjava.lang.String%3B%40fcb045f&oauth_verifier=%5BLjava.lang.String%3B%40598c0112`

url-decoded:

`/login?client_name=[Ljava.lang.String;@221ebc09&oauth_token=[Ljava.lang.String;@fcb045f&oauth_verifier=[Ljava.lang.String;@598c0112`

I'm pretty sure the culprit is in the the following class: `org.jasig.cas.web.flow.CasDefaultFlowUrlHandler` in the method:

```
    @Override
    public String createFlowExecutionUrl(final String flowId, final String flowExecutionKey, final HttpServletRequest request) {
        final StringBuilder builder = new StringBuilder();
        builder.append(request.getRequestURI());
        builder.append('?');
        @SuppressWarnings("unchecked")
        final Map<String, Object> flowParams = new LinkedHashMap<String, Object>(request.getParameterMap());
        flowParams.put(this.flowExecutionKeyParameter, flowExecutionKey);
        appendQueryParameters(builder, flowParams, getEncodingScheme(request));
        return builder.toString();
    }
```

Here `request.getParameterMap()` will return a `Map<String, String[]>` which results in the url above. Apparently `org.springframework.webflow.context.servlet.DefaultFlowUrlHandler#appendQueryParameters` wasn't meant to be fed a `Map<String, String[]>` (which I find weird, since you exactly expect it to receive the parameter map from the request). But this doesn't work really well.

I can't seem to build cas, any idea what I'm doing wrong? I don't have a lot of gradle experience...

```
cas 16h master / ./gradlew build --parallel -x test -x javadoc -DskipCheckstyle=true -DskipAspectJ=true -DskipFindbugs=true                                                                                                                                             java-1.8.0_65
Parallel execution is an incubating feature.

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/auke/projects/workspaces/nrc/cas/build.gradle' line: 1017

* What went wrong:
A problem occurred evaluating root project 'cas-server'.
> No such property: java for class: org.gradle.api.java.archives.internal.DefaultManifest

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 5.516 secs
```

Version information:

```
------------------------------------------------------------
Gradle 2.10
------------------------------------------------------------

Build time:   2015-12-21 21:15:04 UTC
Build number: none
Revision:     276bdcded730f53aa8c11b479986aafa58e124a6

Groovy:       2.4.4
Ant:          Apache Ant(TM) version 1.9.3 compiled on December 23 2013
JVM:          1.8.0_65 (Oracle Corporation 25.65-b01)
OS:           Mac OS X 10.10.5 x86_64
```
Apparently I needed to add

```
apply plugin: 'java'
```

in `build.gradle` for it to work. But even after that I got the following error:


```
uild file '/Users/auke/projects/workspaces/nrc/cas/build.gradle' line: 1286

* What went wrong:
A problem occurred evaluating root project 'cas-server'.
> Cannot add task ':javadoc' as a task with that name already exists.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 10.933 secs
```

I commented the :javadoc task out for now, after that I build succesfully. This was a fully up to date clone of Jasig/cas master.

--- a/cas-server-webapp-actions/src/main/java/org/jasig/cas/web/flow/CasDefaultFlowUrlHandler.java
+++ b/cas-server-webapp-actions/src/main/java/org/jasig/cas/web/flow/CasDefaultFlowUrlHandler.java
@@ -4,8 +4,11 @@ import org.springframework.webflow.context.servlet.DefaultFlowUrlHandler;
 import org.springframework.webflow.core.collection.AttributeMap;
 
 import javax.servlet.http.HttpServletRequest;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
 import java.util.LinkedHashMap;
 import java.util.Map;
+import java.util.stream.Stream;
 
 /**
  * Provides special handling for parameters in requests made to the CAS login
@@ -48,12 +51,19 @@ public final class CasDefaultFlowUrlHandler extends DefaultFlowUrlHandler {
     @Override
     public String createFlowExecutionUrl(final String flowId, final String flowExecutionKey, final HttpServletRequest request) {
         final StringBuilder builder = new StringBuilder();
+        final String encoding = getEncodingScheme(request);
         builder.append(request.getRequestURI());
         builder.append('?');
-        @SuppressWarnings("unchecked")
-        final Map<String, Object> flowParams = new LinkedHashMap<>(request.getParameterMap());
-        flowParams.put(this.flowExecutionKeyParameter, flowExecutionKey);
-        appendQueryParameters(builder, flowParams, getEncodingScheme(request));
+
+        final Map<String, String[]> flowParams = new LinkedHashMap<>(request.getParameterMap());
+        flowParams.put(this.flowExecutionKeyParameter, new String[]{flowExecutionKey});
+
+        final String queryString = flowParams.entrySet().stream()
+                .flatMap(entry -> encodeMultiParameter(entry.getKey(), entry.getValue(), encoding))
+                .reduce((param1, param2) -> param1 + "&" + param2)
+                .orElse("");
+
+        builder.append(queryString);
         return builder.toString();
     }
 
@@ -62,5 +72,21 @@ public final class CasDefaultFlowUrlHandler extends DefaultFlowUrlHandler {
         return request.getRequestURI()
             + (request.getQueryString() != null ? '?'
             + request.getQueryString() : "");
+    }
+
+    private Stream<String> encodeMultiParameter(final String key, final String[] values, final String encoding) {
+        return Stream.of(values).map(value -> encodeSingleParameter(key, value, encoding));
+    }
+
+    private String encodeSingleParameter(final String key, final String value, final String encoding) {
+        return urlEncode(key, encoding) + "=" + urlEncode(value, encoding);
+    }
+
+    private String urlEncode(final String value, final String encodingScheme) {
+        try {
+            return URLEncoder.encode(value, encodingScheme);
+        } catch (final UnsupportedEncodingException e) {
+            throw new IllegalArgumentException("Cannot url encode " + value);
+        }
     }
 }

--- a/cas-server-webapp-actions/src/test/java/org/jasig/cas/web/flow/AllTestsSuite.java
+++ b/cas-server-webapp-actions/src/test/java/org/jasig/cas/web/flow/AllTestsSuite.java
@@ -14,7 +14,7 @@ import org.junit.runners.Suite;
         GenerateServiceTicketActionTests.class, GenericSuccessViewActionTests.class,
         InitialFlowSetupActionTests.class, LogoutActionTests.class,
         SendTicketGrantingTicketActionTests.class, ServiceAuthorizationCheckTests.class,
-        TicketGrantingTicketCheckActionTests.class
+        TicketGrantingTicketCheckActionTests.class, CasDefaultFlowUrlHandlerTests.class
 })
 public class AllTestsSuite {
 }

--- a/cas-server-webapp-actions/src/test/java/org/jasig/cas/web/flow/CasDefaultFlowUrlHandlerTests.java
+++ b/cas-server-webapp-actions/src/test/java/org/jasig/cas/web/flow/CasDefaultFlowUrlHandlerTests.java
@@ -1,0 +1,44 @@
+package org.jasig.cas.web.flow;
+
+import org.junit.Test;
+import org.springframework.mock.web.MockHttpServletRequest;
+
+import static org.junit.Assert.*;
+
+/**
+ * @author Auke van Leeuwen
+ * @since 4.2.0
+ */
+public class CasDefaultFlowUrlHandlerTests {
+
+    private CasDefaultFlowUrlHandler urlHandler = new CasDefaultFlowUrlHandler();
+
+    private MockHttpServletRequest request = new MockHttpServletRequest();
+
+    @Test
+    public void verifyCreateFlowExecutionUrlWithSingleValuedAttributes() throws Exception {
+        setupRequest("/cas", "/app", "/foo");
+        request.setParameter("bar", "baz");
+        request.setParameter("qux", "quux");
+        final String url = urlHandler.createFlowExecutionUrl("foo", "12345", request);
+
+        assertEquals("/cas/app/foo?bar=baz&qux=quux&execution=12345", url);
+    }
+
+    @Test
+    public void verifyCreateFlowExecutionUrlWithMultiValuedAttributes() throws Exception {
+        setupRequest("/cas", "/app", "/foo");
+        request.setParameter("bar", new String[]{"baz1", "baz2"});
+        request.setParameter("qux", "quux");
+        final String url = urlHandler.createFlowExecutionUrl("foo", "12345", request);
+
+        assertEquals("/cas/app/foo?bar=baz1&bar=baz2&qux=quux&execution=12345", url);
+    }
+
+    private void setupRequest(final String contextPath, final String servletPath, final String pathInfo) {
+        request.setContextPath(contextPath);
+        request.setServletPath(servletPath);
+        request.setPathInfo(pathInfo);
+        request.setRequestURI(contextPath + servletPath + pathInfo);
+    }
+}

